TbT data measurement & analysis classes

vellosok75 opened the pull request:
WORK IN PROGRESS: module for measuring and analyzing linear optics from turn-by-turn BPM acquisitions.

The measurement script has been tested online and works well. Only linear optics analysis is available for now (beta beating and phase advance errors), as probed by either sinusoidal fitting, PCA or ICA. Equilibrium parameters fitting (energy spread and emittances) is not yet implemented.

vellosok75 committed:
COMMIS.ENH: initial draft of TbT measurement class

vellosok75 committed:
COMMISS.TBT.ENH: add harmonic analysis methods

murilobalves commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,209 @@
+class TbTDataParams(_AcqBPMsSignalsParams):
+    """."""
+
+    def __init__(self, hkick=None, vkick=None):
```
Allowing inputs is not common in our Params classes. Additionally, this is a potential bug. If the user don't set this optional argument, hkick and vkick will be None and then `prepare_pingers` will set None to the pingers strength, which I think will lead to an error. 

A possible option is that you can initialize `self._hkick = None` and at some point check it is still None, do nothing with the pingers.

vellosok75 replied:
Thanks! Hadn't thought about this possibility. Removed these properties as inputs and added checkings to avoid the  None type to be set as the magnets strengths.

vellosok75 committed:
COMMIS.TBT.BUG: add None type exceptions in str method

vellosok75 committed:
COMMISS.TBT.BUG: fix MeasTbTData initialization

murilobalves commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,209 @@
+        pingh, pingv = self.devices['pingh'], self.devices['pingv']
+        params = self.params
+        hkick, vkick = params['hkick'], params['vkick']
+        pingh.strength = hkick / 1e3   # [urad]
+        pingv.strength = vkick / 1e3   # [urad]
```
It is good to keep our measurement classes as transparent as possible: before the measurement, store the values of parameters that will be changed, change it for the measurement, and then restore the initial state.

xresende replied:
Is it possible for the measurement to be interrupted ? If yes, @vellosok75 did you implement restoration in this situation, apart from restoration at completion ?

murilobalves commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,217 @@
+        currinfo = self.devices['currinfo']
+        self.prepare_pingers()
+        current_before = currinfo.current()
+        self.acquire_data()
```
Maybe just leave a comment briefly describing important data that is stored here from `AcqBPMsSignals`. I would suggest to measure the tunes but then I saw that is already measured in the parent class.

murilobalves commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,217 @@
+from .meas_bpms_signals import AcqBPMsSignals as _AcqBPMsSignals, \
+    AcqBPMsSignalsParams as _AcqBPMsSignalsParams
+
+class TbTDataAnalysis(_AcqBPMsSignals):
```
We typically write the classes following the workflow: Params -> Measurement -> Analysis. It is just a matter of organization, it doesn't affect how the classes work.

vellosok75 replied:
Thanks for the feedback, Murilo. I hadn't yet checked out the acquisition script you mentioned friday. I did it this morning and brought some patterns from there, as well as working on you other suggestions. Thanks!

vellosok75 committed:
COMMISS.TBT.ENH: add restoring strengths methods, remove input to params, and add checking of strengths

vellosok75 committed:
COMMISS.TBT.ENH: change ordering to params, meas, analysis

vellosok75 committed:
COMMISS.TBT.WIP: trying to create magnets timing objects & minor formating fixes

vellosok75 committed:
COMMISS.TBT.ENH: add magnets timing properties

vellosok75 committed:
COMMISS.TBT.ENH: remove unecessary setters for properties

vellosok75 committed:
COMMIS.TBT.ENH: correct initialization of derived classes, remove repeated initialization of parameters

vellosok75 committed:
COMMIS.TBT.ENH: remove fname property from meas class

vellosok75 committed:
COMMISS.TBT.BUG: fix `get_timing_state` s return

vellosok75 committed:
COMMISS.TBT.ENH: keep standard `orbx` & `orby` keys

vellosok75 committed:
COMMISS.TBT.ENH: minor changes in `get_default_fname`

vellosok75 committed:
COMMISS.TBT.ENH: simplify magnets preparation during meas

vellosok75 committed:
COMMIS.ACQBPMS.ENH: remove `do_pulse_evg` and related function

vellosok75 committed:
COMMIS.ACQBPMS.ENH: add timing state to saved data during acquisition

vellosok75 committed:
COMMIS.TBT.ENH: remove redundant function

vellosok75 committed:
COMMISS.TBT.ENH: add selection of magnets to pulse & stregnths setting temporization

vellosok75 committed:
COMMISS.TBT.ENH: check for valid `pingers2kick' settings

fernandohds564 commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,310 @@
+        tm = self.data['timestamp']
+        fmt = '%Y-%m-%d-%H-%M-%S'
+        tmstp = _datetime.datetime.fromtimestamp(tm).strftime(fmt)
+        stg = f'hkick_{int(round(hkick)):3d}_vkick_{int(round(vkick)):3d}'
```
Add information regarding which kicks were kicked.

vellosok75 replied:
This info will be conveyed by the `"hkick"` or `"vkick"` strings and their corresponding values in the default fname. If only one of them is present, then only the corresponding magnet has pulsed. If both are present, both magnets pulsed.

fernandohds564 replied:
I liked the idea of embedding this info in the `hkick` and `vkick` arguments, but maybe it would be better to always keep both on the name. The kicker that didn't pulse could have a different value, though. I was thinking on something like `none` or `nopulse` or `standby` or `inactive` or something else. From these suggestions, I think I prefer `inactive`, but I think you will come up with something better :smile:

fernandohds564 commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,310 @@
+        self._fname = val
+
+    @property
+    def trajx(self):
```
you should get all these properties from the data loaded from file. And remember to return a copy of them, to guarantee data integrity.

fernandohds564 commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,310 @@
+        self._ispost_mortem = False
+
+        self._fname = filename
+        self._trajx, self._trajy = None, None
```
you don't need these properties. See comment below.

fernandohds564 commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,310 @@
+        # load if fname, load method
+
+    @property
+    def fname(self):
```
As we talked offline, I don't think this property is necessary.

vellosok75 replied:
I'll be keeping it for now, following the same patterns of the `OrbitAnalysis` class

vellosok75 commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,310 @@
+            pingv.strength = vkick
+        else:
+            pingv.strength = self.params.vkick / 1e6
+        _time.sleep(0.5)
```
I discussed this with @murilobalves  and @xresende  yesterday. From what I understood, monitors for for pulsed magnets can vary significantly from the set-point values and using them may not be very reliable. Additionally, the `PowerSupplyPU` class has only voltage monitors currently, not strength monitors. 

Do you have any suggestions? Do you think it's worthwhile using monitors and convert voltages to strengths?

vellosok75 replied:
Sorry, I overlooked the `set_strength` function from `PSDev`. Thanks for bringing it up.

vellosok75 committed:
COMMISS.TBT.ENH: handle `None` type in `pingers2kick` & improve the logic of magnets strengths setting

vellosok75 committed:
COMMIS.TBT.ENH: fix formatting

vellosok75 committed:
COMMISS.TBT.ENH: add prefix for kicked beam data, the acq rate and info about which  magnets were pulsed in `get_default_fname`

vellosok75 committed:
COMMIS.TBT.ENH: remove unecessary initializations and properties.

vellosok75 committed:
COMMISS.TBT.BUG: fix pingh trigger device creation

vellosok75 committed:
COMMISS.TBT.ENH: wait magnets ramp during strength setting

vellosok75 committed:
COMMISS.TBT.ENH: formatting fix

vellosok75 commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,310 @@
+        tmstp = _datetime.datetime.fromtimestamp(tm).strftime(fmt)
+        stg = f'hkick_{int(round(hkick)):3d}_vkick_{int(round(vkick)):3d}'
+        stg += f'_urad_{tmstp}'
+        return stg
```
Thought of "kicked_data", but I'm not very happy with it. If you guys have any other suggestions let me know.

vellosok75 committed:
COMMIS.ACQBPM.BUG:  update EVG events after event setting

vellosok75 committed:
COMMISS.ACQBPM.ENH: signal user to trigger timing event.

vellosok75 committed:
COMMISS.TBT.ENH: `magnets_wait_time` -> `magnets_timeout`.

vellosok75 committed:
COMMISS.TBT.ENH: fix magnets waiting time, return flag indicating scucess or failure

vellosok75 committed:
COMMISS.TBT.ENH: improve handling of `pingers2kick`

vellosok75 committed:
COMMIS.TBT.ENH: improve checking and waiting the setting of magnets strengths and measurement workflow

vellosok75 committed:
COMMISS.TBT.BUG: create pingers devices regardless of `pingers2kick`

vellosok75 committed:
COMMISS.TBT.ENH: keep both kicks info in naming system & flag `"inactive"` when not pulsing.

vellosok75 committed:
COMMISS.TBT.ENH: add `magnets_timeout` property in params & improve redability of `__str__` method

vellosok75 committed:
COMMISS.TBT.BUG: prevent attempt at rounding a  `None` type

vellosok75 committed:
COMMISS.TBT.ENH: handle also `"vh"` as keyword for `pingers2kick`

vellosok75 committed:
COMMISS.TBT.ENH: linear fit for fourier components instead of raw-data projection for amplitude guess

vellosok75 committed:
COMMISS.TBT.ENH: "vectorize" harmonic model fitting

vellosok75 commented on apsuite/commisslib/measure_tbt_data.py:
```diff
@@ -0,0 +1,314 @@
+        init_magnets_strength = self.get_magnets_strength()
+        current_before = currinfo.current()
+        self.prepare_timing()
+        self.set_magnets_strength()  # gets strengths from params
```
Thanks, Fernando. I made some changes in the measurement flow, magnets settings and waiting times. Could you take a look again and let me know if you have any other suggestions?

vellosok75 committed:
COMMISS.TBT.ENH: fix  turning magnets on/off

vellosok75 committed:
COMMISS.TBT.ENH: magnets state as a dict with pwrstate and pulsestate

vellosok75 committed:
COMMISS.TBT.ENH: do nothing if strengths are `None` & prepare magnets state compatible with an an optioonal `dict` state arg

vellosok75 committed:
COMMISS.TBT.ENH: include magnets state & strengths to `get_data` function

vellosok75 committed:
COMMISS.TBT.BUG: fixes during online tests, change to mrad.

vellosok75 committed:
COMMISS.TBT.ENH: add analysis class properties and load method

vellosok75 committed:
COMMISS.TBT.ENH: add plotting method to analysis class

vellosok75 committed:
COMMISS.TBT.ENH: add plotting method for beta-beat and phase-errors & method to get the model betas and phases

vellosok75 committed:
COMMISS.TBT.ENH: fix formatting

vellosok75 committed:
COMMISS.TBT.ENH: add harmonic analysis (fitting)

vellosok75 committed:
COMMISS.TBT.ENH: add PCA analysis

vellosok75 committed:
COMMISS.TBT.ENH: do not change manges pwr_state

vellosok75 committed:
COMMISS.TBT.ENH: do not attempt setting mags and doing the meas if `prepare_timing` fails

vellosok75 committed:
TBT.COMMISS.ENH: include the possibility of failing at restoring magnets state

vellosok75 committed:
COMMISS.TBT.BUG: fix default naming to deal with `None`-type strengths

vellosok75 committed:
COMMISS.TBT.BUG: fix restore strengths

vellosok75 committed:
COMMISS.TBT.ANLY: default chromaticity if `chroms == None`

vellosok75 committed:
COMMISS.TBT.MEAS: minor formating fixes

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: wait magnets mon instead of rb

vellosok75 committed:
COMMISS.TBT.ENH: minor improvements in PCA analysis

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: include magnets strengths in magnets state; unify setting of strenghts and states

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: fixes during last online tests

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: improve plot methods

vellosok75 committed:
undo changes in `phase_space.py` committed & pushed by accident

vellosok75 committed:
again to undo stuff done with `phase_space.py`

vellosok75 committed:
COMMISS.TBT.ENH: remove RC params configs

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: rename prorperty `fitting_optics` to `fitting_data`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: plot phase advance and phase advance errors instead of absolute phase at BPMs

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: refactor PCA analysis to share workflow with ICA - whiten source signals, variance contained in mixing matrix

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add ICA analysis with sklearn's FastICA

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: update default params

vellosok75 committed:
COMMISS.AcqBPMS.ENH: get BPM trigger event & mode as strings

vellosok75 committed:
COMMIS.TBT.MEAS.ENH: get pingers trigger event source as strings

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: fixes in default filname

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: fixes in analysis str method

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add default traj turns selection for linear optics fitting

vellosok75 committed:
COMMISS.TbT.ANLY.ENH: fix saving of fitting_data only after loops finish & add some comments

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add reference to docstring

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: fix typo in ICA

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add PCA  & ICA docstrings

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: save also pingers voltage

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add option to compare fit & acquision in `plot_trajs`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: save also fitting residue in `fitting_data`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: plot also residue in `plot_trajs`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add `plot_fit_comparison` & `plot_rms_residue`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: update and add some docstrings

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: fixes fit comparison in `plot_trajs`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: fix tune setting in `_get_nominal_optics`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: save also amplitudes in `harmonic_analysis`

vellosok75 committed:
COMMISS.TBT.ENH: add `plot` argument to plot analysis results only when required

vellosok75 committed:
COMMISS.TBT.ENH.MEAS: turn-off pulse before changing magnets strengths

vellosok75 committed:
COMMISS.TBT.ENH: use int part of tunes only to get nominal optics

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add selection of BPMs  & plot relative phase adv. errs instead of abs errs.

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add `plot_trajs_spectrum`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH:  add spatio-temporal modes analysis

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: add `restore_init_state` option & pingers kicks scaling factors

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: automatically update `trajs_turn_slice`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: handle missing data in `load_and_apply`

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add error bars for PCA analysis

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: set unwrap discont to pi

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: add PCA analysis for stacked xy data

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: update default chromaticity & add TODOs

vellosok75 committed:
COMMISS.TBT.BUG: spec & tunes calculation in stackedxy PCA

vellosok75 committed:
COMMISS.TBT.ANLY.WIP: add SOBI for ICA

vellosok75 committed:
COMMISS.TBT.MEAS.BUG: normalize by scaling factor when getting pingers' strengths

vellosok75 committed:
COMMISS.TBT,BPMS_SIGNAlS.MNT: general formating

vellosok75 committed:
COMMISS.TBT.WIP: migrating meas class to a threaded meas

vellosok75 committed:
COMMISS.TBT:WIP: improve `_do_measurement` flow

vellosok75 committed:
COMMISS.TBT.WIP: improve `set_magnets_strength`'s bad pinger logic

vellosok75 committed:
COMMISS.TBT.WIP: `_set_magnets_strengths` improvements

vellosok75 committed:
COMMISS.TBT.WIP: prevent changing strengths if pinger not used

vellosok75 committed:
COMMISS.TBT.MEAS.BUG: fixes & formatting imprvmnts.

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: init states acquired during initialization

vellosok75 committed:
COMMISS.TBT.MEAS.DEV: check if machine was restored

vellosok75 committed:
COMMISS.TBT.ENH: change pingers delay to delay_raw

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: avoid code repetition in `_get_timing_state`

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: avoid code repeteition in `_get_magnets_strength`

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: add `_acquire_data` method

vellosok75 committed:
COMMISS.TBTMEAS.BUG: fixes
- correct initialization w/ 2 parent classes
- get magnets strength only if pwrstate is on. None otherwise
- turn public some methods

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: simplify setting strengths in `set_magnets_satate`

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: fixes in `set_magnets_strength`

- no wait time during setting of strengths
- introduce `mags_strength_rtol` param

vellosok75 committed:
COMMISS.TBT.MEAS.BUG: use the `_magnets_ok` flag when changing mags

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: move data saving to `acquire_data` & `get_data` only

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: automatically handle `pingers2kick`

- changes `pingers2kick` according to the kick values set

fernandohds564 committed:
COMMISS.TbT: Fix some bugs.

vellosok75 committed:
COMMISS.TbT.MEAS.ENH: some improvements

- `pingers2kick` set by user manually, undoing the second-to-last commit;
- default kicks to 0;
- stop using pingers calibrations during meas
- pingers calibration factors now class constants in analysis class
- avoid initialization of `SOFBFactory` during params & meas classes instantiation

vellosok75 committed:
COMMISS.TBT.ENH: improve meas flow checkings

-`set_magnets_state` now exits soon after failing
- flags for the meas flow control are returned by `prepare_timing` and `prepare_magnets`, instead of manipulating a class var
- stopping the meas externally counts as `meas_finished_ok = False`

vellosok75 committed:
COMMISS.ACQBPMS.ENH: restore support for automatic EVG pulses

- from past discussions we agreed automatic pulsing should not be allowed and the user should always tirgger events manually
- however, without this functionality, the code cannot be automated in a loop

this commit undoes changes from commit 021c505e1e459465d7531fbc5c1823220d4121ad

while preserving changes from  9c408633f1791f272b6eb01c9cd6aaf81ff7e150

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: change default parameter initialization (nr points, do_pulse_evg)

vellosok75 committed:
COMMISS.ANLY.TMP: comment unused param for old data compatibility

vellosok75 committed:
Merge branch 'threaded_tbt_meas' of github.com:lnls-fac/apsuite into threaded_tbt_meas

vellosok75 committed:
COMMISS.TBT.BUG: fixes during online tests

- `get_magnets_strengths` return strengths even if pingers are off
- pingers are turned-on on `set_magnets_state` as called by `prepare_magnets` only if indicated in  `pingers2kick`
- bugfixes related to absent dict keys in `state` dict of `set_magnets_state`
- fix access to the magnets strengths in the str method of analysis class

vellosok75 committed:
COMMISS.TBT.ANLY.BUG: avoid operations with None object if orbx, orby are missing.

vellosok75 committed:
COMMISS.ANLY.ENH: change default timescale in `plot_trajs`& minor sty changes

vellosok75 committed:
COMMISS.TBTANLY.ENH: refactor PCA analysis

- PCA is applied both planes always
- `plot_modal_analysis` reuses code from PCA function
- stetic improvements to `plot_modal_analysis`
- now highlighting singular values of each plane, fixes hor. & ver. signals spatio-temp. -modes
- new structure for `linear_optics_analysis`: choose a method and proceed
- this way, PCA function is more modular. It solely calculates PCA analysis stuff.
- still missing propagate this to ICA analysis and harmonic analysis

vellosok75 committed:
COMMISS.AcqBPMs.ENH: changes during online tests

- standardize event related properties: event_mode, event_delay, etc.
- use delay raw instead of delays
- avoid changing `evg.nr_pulse` if not necessary

vellosok75 committed:
COMMISS.TBT.MEAS.ENH: changes during online tests

- default `do_pulse_evg` to False
- init state properties and function to update init states
- propagate changes from AcqBPMsSginals to standardize "event" properties

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: refactor PCA & modal analysis

- generalized framework for modal analysis to be also applied to ICA
- automatic detection of sin/cos modes
- now also storing sin/cos spatial/temporal modes in `pca_data`
- `plot_modal_analysis` now also indicates sin/cos temporal/spatial modes
- WIP: refactor of ICA

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: refactor ICA

- refactor ICA analysis for stacked hor/ver. planes analysis, by default, as in PCA.
- include alternative to view ICA temporal/spatial modes in `plot_modal_analysis`

Obs: ICA via FastICA is failing to separate signals, and this workflow is failing. SOBI is "working" (seems only to perform PCA), but not understood.

vellosok75 committed:
COMMISS.TBT.ANLY.BUG: minor fixes

vellosok75 committed:
TBT.MNT: minor changes

- comment import from unfinished work in mathphys
- dummy docstrings to properties to remove ruff's complaints

vellosok75 committed:
Merge branch 'master' into tbt-data-analysis

vellosok75 committed:
TBT.ANLY.ENH: handle possibly missing dict keys

vellosok75 committed:
COMMISS.TBT.ANLY.ENH: some standardization

- standardize how hormonic analysis is called (include calc_optitcs arg)
- remove SOBI (unfinished work) as ICA method

vellosok75 committed:
COMMISS.TBT..ANLY.ENH: show phase error in plots

- instead of relative phase error

vellosok75 committed:
COMMISS.TBT.ANLY.MNT: plot relevent planes only

- plot only the relevant plane info (hor/ver) in `linear_optics_analysis`
- change ylabel in sum singal plots

fernandohds564 commented:
@vellosok75, what is missing here for merge?

fernandohds564 committed:
Merge branch 'master' into tbt-data-analysis
